pkg/mimirpb/split_test.go

```diff
@@ -0,0 +1,338 @@
+		assert.Equal(t, []*WriteRequest{
+			{
+				Source:                  RULE,
+				SkipLabelNameValidation: true,
+				Timeseries:              []PreallocTimeseries{req.Timeseries[0]},
+			}, {
+				Source:                  RULE,
+				SkipLabelNameValidation: true,
+				Timeseries:              []PreallocTimeseries{req.Timeseries[1]},
+			}, {
+				Source:                  RULE,
+				SkipLabelNameValidation: true,
+				Metadata:                []*MetricMetadata{req.Metadata[0]},
+			}, {
+				Source:                  RULE,
+				SkipLabelNameValidation: true,
+				Metadata:                []*MetricMetadata{req.Metadata[1]},
+			}, {
+				Source:                  RULE,
+				SkipLabelNameValidation: true,
+				Metadata:                []*MetricMetadata{req.Metadata[2]},
+			},
+		}, partials)
+
+		for _, partial := range partials {
+			assert.Greater(t, partial.Size(), limit)
+		}
+	})
+}
+
+func TestSplitWriteRequestByMaxMarshalSize_Fuzzy(t *testing.T) {
+	const numRuns = 1000
+
+	// Randomise the seed but log it in case we need to reproduce the test on failure.
+	seed := time.Now().UnixNano()
+	rnd := rand.New(rand.NewSource(seed))
+	t.Log("random generator seed:", seed)
+
+	for r := 0; r < numRuns; r++ {
+		var (
+			// Gradually increasing request series in each run, up to a limit.
+			numSeries           = rnd.Intn(min(r+1, 100)) + 1
+			numLabelsPerSeries  = rnd.Intn(min(r+1, 100)) + 1
+			numSamplesPerSeries = rnd.Intn(min(r+1, 100)) + 1
+			numMetadata         = rnd.Intn(min(r+1, 100)) + 1
+		)
+
+		req := generateWriteRequest(numSeries, numLabelsPerSeries, numSamplesPerSeries, numMetadata)
+		maxSize := req.Size() / (1 + rnd.Intn(10))
+		partials := SplitWriteRequestByMaxMarshalSize(req, req.Size(), maxSize)
+
+		// Ensure the merge of all partial requests is equal to the original one.
+		merged := &WriteRequest{
+			Timeseries:              []PreallocTimeseries{},
+			Source:                  partials[0].Source,
+			Metadata:                []*MetricMetadata{},
+			SkipLabelNameValidation: partials[0].SkipLabelNameValidation,
+		}
+
+		for _, partial := range partials {
+			merged.Timeseries = append(merged.Timeseries, partial.Timeseries...)
+			merged.Metadata = append(merged.Metadata, partial.Metadata...)
+		}
+
+		assert.Equal(t, req, merged)
+	}
+}
+
+func TestSplitWriteRequestByMaxMarshalSize_WriteRequestHasChanged(t *testing.T) {
+	// Get WriteRequest field names.
+	val := reflect.ValueOf(&WriteRequest{})
+	typ := val.Type()
+
+	if typ.Kind() == reflect.Ptr {
+		typ = typ.Elem()
+	}
+
+	var fieldNames []string
+	for i := 0; i < typ.NumField(); i++ {
+		fieldNames = append(fieldNames, typ.Field(i).Name)
+	}
+
+	// If the fields of WriteRequest haven't changed, then you will probably need to modify
+	// the SplitWriteRequestByMaxMarshalSize() implementation accordingly!
+	assert.ElementsMatch(t, []string{"Timeseries", "Source", "Metadata", "SkipLabelNameValidation"}, fieldNames)
```

Comment:
Nice check :)

Reply:
It's the first piece of code ever I asked to write to chatgpt.

Reply:
> If the fields of WriteRequest haven't changed

I just realised there's a typo. I wanted to say..."If the fields of WriteRequest HAVE changed".

Reply:
Fixed in https://github.com/grafana/mimir/pull/8077/commits/5126f036773a3ab37ecad5ee5d3654ffd061bdb5